core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -741,6 +758,29 @@ object Doc {
   def split(str: String, pat: Regex = Doc.splitWhitespace, sep: Doc = Doc.lineOrSpace): Doc =
     foldDocs(pat.pattern.split(str, -1).map(Doc.text))((x, y) => x + (sep + y))
 
+  private[paiges] def normalize(d: Doc): Doc = {
+    def concat(a: Doc, b: Doc): Doc = {
+      @tailrec def loop(h: Doc, xs: List[Doc], ys: List[Doc]): Doc = h match {
+        case Concat(a, b) => loop(a, b :: xs, ys)
+        case _ => xs match {
+          case Nil => ys.foldLeft(h){ case (a, b) => Concat(b, a) }
+          case x :: xs => loop(x, xs, h :: ys)
+        }
+      }
+      loop(a, b :: Nil, Nil)
+    }
+    def loop(d: Doc): Doc = d match {
+      case Concat(a, b) => concat(loop(a), loop(b))
```

Comment:
why not `case c@Concat(_, _) => loop(c)` here?

Reply:
I wanted to try to enforce other invariants like no `FlatAlt`s on the left of a `Union` below `Align` and `Nest` for example.  The real intention of `normalize` is that the unflattened docs bubble up all over during `fill` and essentially every invariant and test breaks without putting it in some sensible normal form, like the 2-right-associativity you coined.  This particular one may be too expensive.  I'm happy to hear other ideas.  I just went with some basic ones about associativity and no junk on the left of `Union`s.